### PR TITLE
Prevent last checkpoint being deleted after resumed training with changed `dirpath`

### PR DESCRIPTION
**Pull request by krshrimali**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/11379. Also addresses https://github.com/PyTorchLightning/pytorch-lightning/pull/12045#issuecomment-1051290390.

### Does your PR introduce any breaking changes? If yes, please list them.

Yes, if the checkpoint's `dirpath` has changed on resumed training, `last_model_path` won't be reloaded. In the PR before (https://github.com/PyTorchLightning/pytorch-lightning/pull/12045), we stopped reloading `best_model_score, kth_best_model_path, kth_value, best_k_models`.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @awaelchli @ninginthecloud @jjenniferdai @rohitgr7

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -646,6 +646,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Fixed
 
+- Fixed an issue where `ModelCheckpoint` could delete last checkpoint from the old directory when `dirpath` has changed during resumed training ([]())
```
Comment (krshrimali) [inline CHANGELOG.md:649]:
```suggestion
- Fixed an issue where `ModelCheckpoint` could delete last checkpoint from the old directory when `dirpath` has changed during resumed training ([#12225](https://github.com/PyTorchLightning/pytorch-lightning/pull/12225))
```

File: tests/checkpointing/test_model_checkpoint.py [hunk 2]
```diff
@@ -1217,8 +1217,8 @@ def make_assertions(cb_restore, written_ckpt):
             "kth_best_model_path": False,
             "kth_value": False,
             "best_k_models": False,
+            "last_model_path": False,
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:1220]: Can you add a test? You can mimic the behaviour shown by Adrian in https://github.com/PyTorchLightning/pytorch-lightning/pull/12045#issuecomment-1051290390

Comment (krshrimali) [reply tests/checkpointing/test_model_checkpoint.py:1220]: Yes, that makes sense. Done! Please let me know if it looks good. :)

File: tests/checkpointing/test_model_checkpoint.py [hunk 3]
```diff
@@ -1270,6 +1270,77 @@ def on_load_checkpoint(self, *args, **kwargs):
...
+        def __init__(self):
+            super().__init__()
+            self.layer = torch.nn.Linear(8, 2)
+
+        def forward(self, x):
+            return self.layer(x)
+
+        def training_step(self, batch, batch_idx):
+            loss = self(batch).sum()
+            self.log("latest_is_best", self.global_step)
```
Comment (krshrimali) [inline tests/checkpointing/test_model_checkpoint.py:1291]: We are logging `latest_is_best`, hence not importing it from existing helper functions.

File: tests/checkpointing/test_model_checkpoint.py [hunk 4]
```diff
@@ -1270,6 +1270,77 @@ def on_load_checkpoint(self, *args, **kwargs):
     make_assertions(cb_restore, written_ckpt)
 
 
+def test_resume_training_preserves_old_ckpt_last(tmpdir):
+    # This test ensures that the last checkpoint when saved is not deleted from the previous folder
+    # (when training is resumed from the old checkpoint)
+
+    from torch.utils.data import DataLoader
+
+    from pytorch_lightning import LightningModule
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:1279]: No need for local imports here, you can move these to the top of the file

File: tests/checkpointing/test_model_checkpoint.py [hunk 5]
```diff
@@ -1270,6 +1270,77 @@ def on_load_checkpoint(self, *args, **kwargs):
     make_assertions(cb_restore, written_ckpt)
 
 
+def test_resume_training_preserves_old_ckpt_last(tmpdir):
+    # This test ensures that the last checkpoint when saved is not deleted from the previous folder
+    # (when training is resumed from the old checkpoint)
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:1275]: This should be a triple quoted docstring

```
"""bla bla bla"""
```

File: tests/checkpointing/test_model_checkpoint.py [hunk 6]
```diff
@@ -1270,6 +1270,77 @@ def on_load_checkpoint(self, *args, **kwargs):
...
+
+    from torch.utils.data import DataLoader
+
+    from pytorch_lightning import LightningModule
+
+    class BoringModel(LightningModule):
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:1281]: You can subclass the BoringModel instead and avoid overriding all other methods thatn `training_step`

```suggestion
    class MyModel(BoringModel):
```

File: tests/checkpointing/test_model_checkpoint.py (hunk 3 above)
Comment (carmocca) [reply tests/checkpointing/test_model_checkpoint.py:1291]: What do you mean?

File: tests/checkpointing/test_model_checkpoint.py [hunk 7]
```diff
@@ -1270,6 +1270,77 @@ def on_load_checkpoint(self, *args, **kwargs):
...
+
+        def training_step(self, batch, batch_idx):
+            loss = self(batch).sum()
+            self.log("latest_is_best", self.global_step)
+            return {"loss": loss}
+
+        def configure_optimizers(self):
+            return torch.optim.SGD(self.layer.parameters(), lr=0.1)
+
+    train_data = DataLoader(RandomDataset(8, 16), batch_size=2)
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:1297]: No need to define the dataloader if we subclass the `BoringModel`

```suggestion
```

File: tests/checkpointing/test_model_checkpoint.py [hunk 8]
```diff
@@ -1270,6 +1270,77 @@ def on_load_checkpoint(self, *args, **kwargs):
...
+        dirpath=tmpdir,
+        monitor="latest_is_best",
+        mode="max",
+        save_last=True,
+        filename="{epoch}-{step}-{latest_is_best}",
+        save_top_k=3,
+        every_n_train_steps=1,
+    )
+    trainer = Trainer(
+        default_root_dir=os.getcwd(),
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:1310]:
```suggestion
        default_root_dir=tmpdir,
```

File: tests/checkpointing/test_model_checkpoint.py [hunk 9]
```diff
@@ -1270,6 +1270,77 @@ def on_load_checkpoint(self, *args, **kwargs):
...
+    new_ckpt = ModelCheckpoint(
+        dirpath=str(tmpdir / "after-reload"),
+        monitor="latest_is_best",
+        mode="max",
+        save_last=True,
+        filename="{epoch}-{step}-{latest_is_best}",
+        save_top_k=3,
+        every_n_train_steps=1,
+    )
+    # Training it for 2 epochs for extra surity, that nothing gets deleted after multiple epochs
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:1330]:
```suggestion
    # Training it for 2 epochs for extra surety, that nothing gets deleted after multiple epochs
```

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 10]
```diff
@@ -623,7 +623,7 @@ def _get_metric_interpolated_filepath_name(
 
     def _monitor_candidates(self, trainer: "pl.Trainer", epoch: int, step: int) -> Dict[str, _METRIC]:
         monitor_candidates = deepcopy(trainer.callback_metrics)
-        monitor_candidates.update(epoch=epoch, step=step)
+        monitor_candidates.update(epoch=torch.tensor(epoch), step=torch.tensor(step))
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/model_checkpoint.py:626]: was this is a bug before?

Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:626]: Yes

File: tests/checkpointing/test_model_checkpoint.py [hunk 11]
```diff
@@ -1247,6 +1247,40 @@ def on_load_checkpoint(self, *args, **kwargs):
...
+        "monitor": "step",
+        "mode": "max",
+        "save_last": True,
+        "save_top_k": 2,
+        "every_n_train_steps": 1,
+    }
+    trainer = Trainer(**trainer_kwargs, callbacks=ModelCheckpoint(**mc_kwargs))
+    trainer.fit(model)
+    # Make sure that the last checkpoint file exists in the dirpath passed (`tmpdir`)
+    assert sorted(os.listdir(tmpdir / "checkpoints")) == sorted(["last.ckpt", "step=2.ckpt", "step=3.ckpt"])
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:1273]: Order doesn't matter, could use a set

```suggestion
    assert set(os.listdir(tmpdir / "checkpoints")) == {"last.ckpt", "step=2.ckpt", "step=3.ckpt"}
```

File: tests/checkpointing/test_model_checkpoint.py [hunk 12]
```diff
@@ -1194,8 +1194,8 @@ def make_assertions(cb_restore, written_ckpt):
             "kth_best_model_path": False,
             "kth_value": False,
             "best_k_models": False,
+            "last_model_path": False,
```
Comment (rohitgr7) [inline tests/checkpointing/test_model_checkpoint.py:1197]: do we need the other additional test? wouldn't this one satisfy the requirement for the requirement we need?

**Review by rohitgr7 — APPROVED** [review]

File: tests/checkpointing/test_model_checkpoint.py (hunk 12 above)
Comment (carmocca) [reply tests/checkpointing/test_model_checkpoint.py:1197]: This test is more contained and just checks that the state is not restored.

But we could have a bug in the "save last" saving logic after reload which should be covered by the new one.

So you are correct depending on how lenient we want to be with our tests.

**Review by Borda — APPROVED** [review]
lgtm

**Comment by krshrimali** [comment]
Reviews: will fix merge conflicts in an hour, and hopefully we can merge then if all is good. :)
